### DISCUSSION TITLE
The context/environment is not secure

### DISCUSSION THREAD

**Discussion by madtag** [soofstad/react-oauth2-pkce#83] [Q&A] [answered]
Hi all,

I am trying to use this module to authenticate on keycloak.  After solving cors issues I am not getting the error below.  Tried on Chrome just in case and got the same error. Any ideas on how to solve this?
![image](https://user-images.githubusercontent.com/5601734/234827480-eba716ed-b398-43cb-8d61-6dff9a9d9c58.png)

**Comment by sebastianvitterso** [comment, marked as answer]
Hi, @madtag!
This error is thrown by [pkceUtils.ts:24-30](https://github.com/soofstad/react-oauth2-pkce/blob/main/src/pkceUtils.ts#L24-L30). 
It can happen because of an outdated browser, but it's probably rather because you are trying to use it in an "unsecure context" where SubtleCrypto is disabled automatically by the browser (for security reasons). (MDN docs on [`crypto.subtle`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/subtle)) 

Any url that starts with `http://` instead of `https://` (except `localhost`, I think) are considered insecure automatically. 

Two quick ways to check if this is the issue:
1. Tap the address bar in your browser, and copy and paste it. Does it say `http://` at the beginning (and the rest isn't `localhost`) - then that's your issue. 
2. To check whether javascript considers the website to be secure, simply open your dev console and type in `window.isSecureContext`, to see what it returns. If that's false, you need to look at why. (MDN docs on [Secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts)).

To solve this issue, you probably need to host your website using TLS (HTTPS).

Let us know if that wasn't your issue, and we'll have a look!

**Comment by madtag** [comment]
So I checked with window.isSecureContext and you are correct about http being unsecure and https and localhost being secure.  This is the problem now.  using http://localhost:3000 when I try to login I get a CORS error.  If I use http://myIP:3000 I get the context error.

If I run my application using HTTPS=true in .env file for both http://localhost:3000 and https://myIP:3000 i get the error below when trying to login

![image](https://user-images.githubusercontent.com/5601734/234845755-3e34385a-b096-4eda-9b90-37a7773be8e2.png)

**Reply by madtag** [reply]
I am trying to deploy my code on the server now to see what happens.

**Reply by soofstad** [reply]
The "normal" solution to this is to add "http://localhost:3000" as a valid address in the auth provider, to whitelist it for CORS errors.

If you are uncomfortable with that for security reasons, you could create a new client in the auth provider for local development only.

**Reply by madtag** [reply]
![image](https://user-images.githubusercontent.com/5601734/234847858-bb67f915-cc2b-44ab-be71-e8145cae196a.png)

**Reply by soofstad** [reply]
In Keycloak, valid redirect does not update CORS.
You will need to add localhost to "Web origins" as well.
![Screenshot_20230427_133914](https://user-images.githubusercontent.com/11062560/234851293-9c45635a-e952-41da-9acd-ce6100e6b00b.png)

**Reply by madtag** [reply]
This solved the cors issue.  I selected the first comment as the answer to the question because that solved the security issue. Thanks a million mate.

**Reply by soofstad** [reply]
Good stuff! Happy to help. Give the project a star if you haven't already 😉
